### PR TITLE
Dvc 4843 refetch config on message

### PR DESCRIPTION
**Pull request by emmawillis**

### CODE REVIEW

File: android-client-sdk/src/main/java/com/devcycle/sdk/android/api/DVCClient.kt [hunk 1]
```diff
@@ -53,14 +55,28 @@ class DVCClient private constructor(
...
+                    eventSource = EventSource.Builder(Handler(fun (event: String?, messageEvent: MessageEvent?) {
+                        val data = JSONObject(messageEvent?.data)
+                        var type = ""
+                        try {
+                            type = data.get("type") as String
+                        } catch (e: Exception) {}
+
+                        println("!! in message callback: type: $type query: ${messageEvent?.origin?.query}") // TODO remove
+
+                        if (type == "refetchConfig" && messageEvent?.origin?.query?.contains("sse=1") == true) { // TODO not sure if 'origin' is where the query param would show up?
```
Comment (ajwootto) [inline android-client-sdk/src/main/java/com/devcycle/sdk/android/api/DVCClient.kt:76]: this should just check if the type is refetchConfig or the type is not set.

The SSE param is sent by the SDK to the config API during the refetchConfig operation (which should also be happening here)

File: android-client-sdk/src/main/java/com/devcycle/sdk/android/api/DVCClient.kt [hunk 2]
```diff
@@ -53,14 +55,28 @@ class DVCClient private constructor(
...
+                            refetchConfig()
```
Comment (ajwootto) [inline android-client-sdk/src/main/java/com/devcycle/sdk/android/api/DVCClient.kt:77]: also send the `configLastModified` param

File: android-client-sdk/src/main/java/com/devcycle/sdk/android/api/DVCClient.kt [hunk 3]
```diff
@@ -61,8 +64,20 @@ class DVCClient private constructor(
...
+                    var lastModified: Long = 0
+                    try {
+                        type = data.get("type") as String
+                        lastModified = data.get("lastModified") as Long
+                    } catch (e: JSONException) {}
+
+                    if (type == "refetchConfig" || type == "") { // Refetch the config if theres no type
+                        refetchConfig(true, lastModified)
+                    }
+                }), URI(config?.sse?.url)).build()
```
Comment (ajwootto) [inline android-client-sdk/src/main/java/com/devcycle/sdk/android/api/DVCClient.kt:79]: the event source should only be created if this URL exists in the config

File: android-client-sdk/src/main/java/com/devcycle/sdk/android/api/DVCClient.kt [hunk 4]
```diff
@@ -309,9 +324,9 @@ class DVCClient private constructor(
         dvcSharedPrefs.save(user, DVCSharedPrefs.UserKey)
     }
 
-    private suspend fun fetchConfig(user: User) {
+    private suspend fun fetchConfig(user: User, sse: Boolean? = false, lastModified: Long? = 0) {
```
Comment (ajwootto) [inline android-client-sdk/src/main/java/com/devcycle/sdk/android/api/DVCClient.kt:327]: lastModified shouldnt default to 0

File: android-client-sdk/src/main/java/com/devcycle/sdk/android/api/DVCClient.kt [hunk 5]
```diff
@@ -61,8 +64,20 @@ class DVCClient private constructor(
             try {
                 fetchConfig(user)
                 isInitialized.set(true)
+                eventSource = EventSource.Builder(Handler(fun (event: String?, messageEvent: MessageEvent?) {
+                    val data = JSONObject(messageEvent?.data)
+                    var type = ""
+                    var lastModified: Long = 0
```
[the remark this answers is not in the document]
Comment (emmawillis) [reply android-client-sdk/src/main/java/com/devcycle/sdk/android/api/DVCClient.kt]: fixed!

File: android-client-sdk/src/main/java/com/devcycle/sdk/android/api/Request.kt [hunk 6]
```diff
@@ -62,6 +64,12 @@ internal class Request constructor(envKey: String, apiBaseUrl: String, eventsBas
         if (enableEdgeDB) {
             map["enableEdgeDB"] = "true"
         }
+        if(sse == true) {
+            map["sse"] = "true"
+        }
+        if(lastModified != null) {
+            map["lastModified"] = "$lastModified"
```
Comment (ajwootto) [inline android-client-sdk/src/main/java/com/devcycle/sdk/android/api/Request.kt:71]: the query param should be `sseLastModified`

File: android-client-sdk/src/main/java/com/devcycle/sdk/android/eventsource/EventHandler.kt [hunk 7]
```diff
@@ -3,7 +3,7 @@ package com.devcycle.sdk.android.eventsource
 import timber.log.Timber
 
 class Handler(
-    private var onMessageHandler: () -> Any
+    private var onMessageHandler: (event: String?, messageEvent: MessageEvent?) -> Any
```
Comment (ajwootto) [inline android-client-sdk/src/main/java/com/devcycle/sdk/android/eventsource/EventHandler.kt:6]: will the handler ever care about both of these properties, or does it just want the messageEvent?

Comment (emmawillis) [reply android-client-sdk/src/main/java/com/devcycle/sdk/android/eventsource/EventHandler.kt:6]: good point, the event seems to always be 'message' so i can't see it ever being useful

File: android-client-sdk/src/main/java/com/devcycle/sdk/android/api/DVCClient.kt (hunk 3 above)
Comment (emmawillis) [reply android-client-sdk/src/main/java/com/devcycle/sdk/android/api/DVCClient.kt]: fixed!

File: android-client-sdk/src/main/java/com/devcycle/sdk/android/api/DVCClient.kt [hunk 8]
```diff
@@ -61,9 +64,23 @@ class DVCClient private constructor(
             try {
                 fetchConfig(user)
                 isInitialized.set(true)
-                withContext(Dispatchers.IO){
-                    eventSource = EventSource.Builder(Handler(fun () {}), URI(config?.sse?.url)).build()
-                    eventSource.start()
+                if (config?.sse?.url != null) {
+                    withContext(Dispatchers.IO) {
+                        eventSource = EventSource.Builder(Handler(fun(messageEvent: MessageEvent?) {
+                            val data = JSONObject(messageEvent?.data)
```
Comment (ajwootto) [inline android-client-sdk/src/main/java/com/devcycle/sdk/android/api/DVCClient.kt:70]: should this just early return if the messageEvent is null? Also is that even possible?

Comment (emmawillis) [reply android-client-sdk/src/main/java/com/devcycle/sdk/android/api/DVCClient.kt]: i can just wrap all the other logic in an if statement checking that it's not null, so it does nothing if the message event is null for some reason

**Pull request merged**